Onnx trt

yinghai opened the pull request:
onnx model ->trt op flow
@bddppq

Commit:
Add support to TensorRT Op

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,184 @@
+  }
+  std::cout << std::endl;
+
+  if (nv_dims.nbDims + 1 != c2_dims.size()) {
```
just make it all CAFFE_ENFORCE with appropriate error messages

writing to raw cout is bad

yinghai replied:
I forgot to remove the debug string... Will remove them.

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,184 @@
+  std::unordered_map<std::string, int> outputs;
+  for (int i = 0; i < operator_def.input_size(); ++i) {
+    inputs.emplace(operator_def.input(i), i);
+    std::cout << "Adding Input: " << operator_def.input(i) << std::endl;
```
VLOG

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,184 @@
+    const auto& name = trt_engine_->getBindingName(b);
+    nv_dims_.push_back(trt_engine_->getBindingDimensions(b));
+    if (trt_engine_->bindingIsInput(b)) {
+      std::cout << "Checking TRT input: " << name << std::endl;
```
let's remove cout and do LOG(INFO) / VLOG

yinghai replied:
Yeah, it's debug info and will be removed.

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,184 @@
+    if (trt_engine_->bindingIsInput(b)) {
+      std::cout << "Checking TRT input: " << name << std::endl;
+      const auto it = inputs.find(name);
+      CAFFE_ENFORCE(it != inputs.end(), MakeString("Cannot find trt input: ", name));
```
I wonder whether we should bind inputs by position here. In case somebody rewrites the net after the TRT op is create it'd stop working. In C2 we shouldn't rely on the names of inputs in the graph to do matching, in general

yinghai replied:
Yes, the safest way is to do it in `RunOnDevice`. But I need more info on  

> In C2 we shouldn't rely on the names of inputs in the graph to do matching, in general

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,184 @@
+      N = tensor_dims.front();
+      first = false;
+    } else {
+      CAFFE_ENFORCE(
```
CAFFE_ENFORCE_EQ

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,184 @@
+          N == tensor_dims.front(), "Mismatched batch size in input tensors");
+    }
+  }
+  CAFFE_ENFORCE(N <= batch_size_, "Batch size is too large");
```
CAFFE_ENFORCE_LE

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,184 @@
+  CAFFE_ENFORCE(N <= batch_size_, "Batch size is too large");
+  batch_size_ = N;
+
+  // We need to do the binding at RunOnDevice time because we only know the
```
it's usually a good idea to cache bindings between invocations in case pointers or sizes don't change. But it can be a TODO for now

yinghai replied:
I thought about it but I wasn't sure if the sizes are guaranteed not to change or not. Binding itself doens't cost much. More computation is used in dim size comparison.

dzhulgakov commented on caffe2/python/pybind_state_gpu.cc:
```diff
@@ -66,6 +67,17 @@ void addCUDAGlobalMethods(py::module& m) {
     obj["totalGlobalMem"] = py::cast(prop.totalGlobalMem);
     return obj;
   });
+  m.def(
```
it won't compile if we compile without USE_TENSORRT, right? we probably should create a tiny separate pybind

yinghai replied:
Yes. Plumbing through the flow now. Will change.

dzhulgakov commented on caffe2/trt/tensorrt_tranformer.h:
```diff
@@ -0,0 +1,38 @@
+
+namespace caffe2 {
+
+  class TensorRTTransformer {
```
hm, why is it a class and not a function?

yinghai replied:
Because I plan to add graph cutting in this class too.

dzhulgakov reviewed (commented):
Good stuff! some minor comments here and there

dzhulgakov commented on CMakeLists.txt:
```diff
@@ -54,12 +54,14 @@ option(USE_NNAPI "Use NNAPI" OFF)
 option(USE_NNPACK "Use NNPACK" ON)
 option(USE_NUMA "Use NUMA (only available on Linux)" ON)
 option(USE_OBSERVERS "Use observers module." OFF)
+option(USE_ONNX_TRT "Using ONNX-TensorRT library" OFF)
```
hm - why do we have two switches for it? when I'd use one and not another?

yinghai replied:
This is only one minor case, for inference, we can load the serialized c2 model with trt op. It doesn't require onnx-trt. But the real reason is that I added tensorRT and onnx-trt in two commits so I'd like to tune it during development. :) I will merge them.

dzhulgakov commented on caffe2/onnx/onnx_exporter.cc:
```diff
@@ -364,6 +364,67 @@ std::vector<NodeProto> OnnxExporter::CreateGemmNode(
   return nodes;
 }
+
+void OnnxExporter::InitOpToTensorProto(
```
I think relying on init_net in this case is hacky. Usually caffe2 predictor (e.g. see predictor exporter) takes predict net and workspace and pulls params directly from workspace. We should do the same here as it's more generic and extensible.

And if you have init_net - the way to go to workspace is to just run it.

yinghai replied:
Ok. I'm not aware of the predictor exporter and it's workflow. I did this because our c2 model zoo came in as init/pred net pair and I was biased on that. I agree it's more generic. I'll collect more info on this. Change itself shouldn't be too difficult.

dzhulgakov commented on caffe2/operators/tensorrt_op_trt.h:
```diff
@@ -0,0 +1,53 @@
+  TrtLogger logger_;
+  int batch_size_;
+  std::vector<void*> bindings_;
+  std::vector<std::pair<int, bool>> binding_hints_;
```
add a comment what the bool is? (seems input/output)

dzhulgakov commented on caffe2/trt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,465 @@
+  return all_outputs_vec;
+}
+
+std::vector<ONNX_NAMESPACE::ValueInfoProto> ConvertToValueInfo(
```
prepend ONNX_NAMESPACE everywhere with :: - otherwise it might conflict with caffe2::onnx

dzhulgakov commented on caffe2/trt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,465 @@
+  return r;
+}
+
+void PruneWeights(const NetDef& pred_net, NetDef* init_net) {
```
again, I don't think relying on init_net is good here. Many C2 models are stored in DB + predict_net format.

You might still need to pass the set of params to filter, but the weights themselves are better to live in Workspace

dzhulgakov commented on caffe2/trt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,465 @@
+  LOG(INFO) << "New init_net op size: " << init_net->op_size();
+}
+
+void FillModelInfo(ONNX_NAMESPACE::ModelProto* model) {
```
hm, why it's not in onnx_exporter somewhere?

yinghai replied:
It should.

dzhulgakov commented on caffe2/trt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,465 @@
+  auto importer = InferObject(onnx2trt::createImporter(trt_network.get()));
+  auto status = importer->import(onnx_model_str.data(), onnx_model_str.size(), false);
+  if (status.is_error()) {
+    LOG(ERROR) << "TensorRTTransformer ERROR: " << status.file() << ":"
```
throw?

dzhulgakov commented on caffe2/trt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,465 @@
+
+  auto engine_plan = InferObject(trt_engine->serialize());
+
+  auto* serialized_engine_arg = op.add_arg();
```
there are helper functions for it, like MakeArgument

dzhulgakov commented on caffe2/trt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,465 @@
+  auto* output_size_hints_arg = op.add_arg();
+  auto* output_size_names_arg = op.add_arg();
+  output_size_hints_arg->set_name("output_size_hints");
+  output_size_names_arg->set_name("output_size_names");
```
another option would be to create multiple arguments with dynamic names. It's not perfect but might be less hacky than -1 delimited list

yet another option is a string with json

but I don't have strong preference

yinghai replied:
Why we no have maps in protobuf? :(

dzhulgakov reviewed (commented):
Looks great, some subset of comments. Let's ping more folks to review it thoroughly

Commit:
dpm synchronize

Commit:
Caffe2-onnx exporter (#2248)

* caffe2-onnx frontend

* Remove Python part of the conversion code

* nit

* convert more ops

* Address commmetns

mlappelbaum committed:
Export atomic iter count (#2379)

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Add axis to top_k_op. (#2416)

* Revert update on top_k_op

* Add axis to top_k_op

Add axis to top_k_op

* [auto] Update onnx to a8e4648 - Adjust link flags when built in Windows Debug mode (#647)
https://github.com/onnx/onnx/commit/a8e4648a7de0992768d74e6c43d00a43ee983a76

* [auto] Update onnx to f4acf28 - Remove allowconsumed enforceconsumed from op schema. (#617)
https://github.com/onnx/onnx/commit/f4acf281eff7afe81f29f014edaa87e4c04c01a3

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Initialize cpuinfo in the thread pool

Thread pool called cpuinfo_get_processors_count() without initializing cpuinfo. Only by luck it didn't make Caffe2 single-threaded: threadpool is initialized after NNPACK, and NNPACK initializes cpuinfo itself.

This commit also updates cpuinfo to a version that aborts with a fatal error if its used uninitialized.

* Updated Python Op and Image Pre-Processing Pipeline tutorials && Added CIFAR-10 Part 1 tutorial (#2286)

* Updated Basics tutorial: (1) Added Python 3 support with __future__ statements; (2) Various grammatical/typo fixes and minor refactoring of Markdown

* Added Python 3 support and made minor typo fixes

* Added Python 3 support with future imports, refactored and corrected errors in Markdown, added comments

* Added Python 3 support with future imports, Added use of caffe_translator.py to translate downloaded .caffemodel file to .pb files

* Upgrades to Image Pre-Processing Pipeline tutorial

* Updated Python Op tutorial

* removed markdown with empty links

* Added Part 1 of an end-to-end CIFAR-10 tutorial

* Updated MNIST Dataset and Databases tutorial with python3 support and markdown fixes

* Tweaks to markup, less training iterations

* changed permissions of CIFAR10_Part1; typo corrections in Image_Pre-Processing_Pipeline

* Typo corrections in Multi-GPU Training tutorial

* sync Python_Op py_gen with the IPython notebook

* nit typo correction

* [auto] Update onnx to 5cb999d - Minor cleanups to shape inference (#653)
https://github.com/onnx/onnx/commit/5cb999ddc13efbe9a29346a620834bae6b2c764e

* [auto] Update onnx to ecac1c1 - Merge Rel 1.1.0 branch into master (#657)
https://github.com/onnx/onnx/commit/ecac1c162485b3d6072eff73dbbc79fd5376c8ff

* Strip down onnx to only pb definitions in mobile build (#2426)

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

* Exported AtomicIterOp count

orionr committed:
Move doc generation code into docs/caffe2. (#2435)

yinghai commented:
Closing in favor of formal PR: https://github.com/caffe2/caffe2/pull/2443
